cabal-install/Distribution/Solver/Modular/Solver.hs

```diff
@@ -101,4 +124,87 @@ solve sc cinfo idx pkgConfigDB userPrefs userConstraints userGoals =
                                                   , PackageName "integer-gmp"
                                                   , PackageName "integer-simple"
                                                   ])
-    buildPhase       = addLinking $ buildTree idx (independentGoals sc) userGoals
+    buildPhase       = traceTree "build.json" id
+                     $ addLinking
+                     $ buildTree idx (independentGoals sc) userGoals
+
+-- | Dump solver tree to a file (in debugging mode)
+--
+-- This only does something if the @debug-tracetree@ configure argument was
+-- given; otherwise this is just the identity function.
+traceTree ::
+#ifdef DEBUG_TRACETREE
+  GSimpleTree a =>
+#endif
+     FilePath  -- ^ Output file
+  -> (a -> a)  -- ^ Function to summarize the tree before dumping
```

Comment:
Looks like you're only calling `traceTree` with `id`, so in principle you could get rid of this parameter.